coalib/core/Bear.py

```diff
@@ -0,0 +1,502 @@
+            if cp is not False:
+                error_string += ' ' + cp
+
+            raise RuntimeError(error_string)
+
+    def reset_dependency_results(self):
+        """
+        Resets all dependency results for this bear.
+
+        >>> section = Section('my-section')
+        >>> file_dict = {'file1.txt': ['']}
+        >>> bear = Bear(section, file_dict)
+        >>> dependency_bear = Bear(section, file_dict)
+        >>> bear.add_dependency_results(dependency_bear, [1, 2, 3])
+        >>> bear.dependency_results
+        {<class 'coalib.core.Bear.Bear'>: [1, 2, 3]}
+        >>> bear.reset_dependency_results()
+        >>> bear.dependency_results
+        {}
+        """
+        self._dependency_results = {}
+
+    def add_dependency_results(self, bear, dependency_results):
+        """
+        Adds dependency results to this instance.
+
+        This function is used by the core to add dependency results as they
+        are ready.
+
+        >>> section = Section('my-section')
+        >>> file_dict = {'file1.txt': ['']}
+        >>> bear = Bear(section, file_dict)
+        >>> bear.dependency_results
+        {}
+        >>> dependency_bear = Bear(section, file_dict)
+        >>> bear.add_dependency_results(dependency_bear, [1, 2, 3])
+        >>> bear.dependency_results
+        {<class 'coalib.core.Bear.Bear'>: [1, 2, 3]}
+
+        :param bear:
+            The bear instance to add results for.
+        :param dependency_results:
+            The results to add.
+        """
+        bear_type = type(bear)
+
+        if bear_type not in self._dependency_results:
+            self._dependency_results[bear_type] = list(dependency_results)
+        else:
+            self._dependency_results[bear_type] += dependency_results
+
+    @property
+    def dependency_results(self):
+        """
+        Contains all dependency results.
+
+        This variable gets set during bear execution from the core and can be
+        used from ``analyze``.
+
+        Modifications to the returned dictionary lead to undefined behaviour.
+
+        >>> section = Section('my-section')
+        >>> file_dict = {'file1.txt': ['']}
+        >>> bear = Bear(section, file_dict)
+        >>> bear.dependency_results
+        {}
+        >>> dependency_bear = Bear(section, file_dict)
+        >>> bear.add_dependency_results(dependency_bear, [1, 2, 3])
+        >>> bear.dependency_results
+        {<class 'coalib.core.Bear.Bear'>: [1, 2, 3]}
+
+        :return:
+            A dictionary with bear-types as keys and their results received.
+        """
+        return self._dependency_results
+
+    @classmethod
+    def get_metadata(cls):
+        """
+        :return:
+            Metadata for the ``analyze`` function extracted from its signature.
+            Excludes parameter ``self``.
+        """
+        return FunctionMetadata.from_function(
+            cls.analyze,
+            omit={'self'})
+
+    # FIXME Make this a @classproperty.
```

Comment:
but we have to integrate the core into the current system, and this one can be easily changed after integration -> smaller commits + less problems (imo)